Support for --silent flag in api

AliabbasMerchant opened the pull request:
## Summary

closes #1188

AliabbasMerchant committed:
--silent flag in api

AliabbasMerchant committed:
Added Test for --silent flag in api

AliabbasMerchant committed:
Skip printing headers when --silent in api

mislav commented on pkg/cmd/api/api.go:
```diff
@@ -178,6 +180,10 @@ func apiRun(opts *ApiOptions) error {
 	}
 
+	if opts.Silent {
+		opts.IO.Out = ioutil.Discard
```
This is a neat trick! But, since parsing response JSON will take some processing time and will ultimately be discarded, could you instead skip the body processing logic in "silent" mode and exit from the command early?

mislav replied:
On the other hand, I just realized that the response body needs to be first consumed in order for our response error processing logic to work. So, maybe discard my above comment please, as following my suggestion might lead to too many code changes. I'll leave it to you!

AliabbasMerchant replied:
I think this Discard option is better than some complex logic. I agree with you on the error handling part. Even if I put the custom logic and all, it will become too bloated, and will cause problems down the line also, in case we need to add/change some feature/working.

mislav commented on pkg/cmd/api/api.go:
```diff
@@ -212,7 +218,7 @@ func apiRun(opts *ApiOptions) error {
 
 func processResponse(resp *http.Response, opts *ApiOptions) (endCursor string, err error) {
-	if opts.ShowResponseHeaders {
+	if opts.ShowResponseHeaders && !opts.Silent {
```
I don't think we need to change this conditional. In fact, if someone requested response headers, let's display them even in `--silent` mode. Maybe there is some use-case for that, and I think it's safe to allow because both `-i` and `--silent` flags are explicitly opt-in.

AliabbasMerchant replied:
Right. Makes sense. I will remove that.

mislav commented on pkg/cmd/api/api_test.go:
```diff
@@ -425,6 +449,38 @@ func Test_apiRun_paginationGraphQL(t *testing.T) {
 	assert.Equal(t, "PAGE1_END", endCursor)
 }
 
+func Test_apiRun_silent(t *testing.T) {
```
Instead of adding a separate test function, could you add the `Silent: true` mode as another test case in `Test_apiRun` please?

AliabbasMerchant replied:
Oh. Right. I will do that!

mislav commented on pkg/cmd/api/api.go:
```diff
@@ -134,6 +135,7 @@ original query accepts an '$endCursor: String' variable and that it fetches the
 	cmd.Flags().BoolVarP(&opts.ShowResponseHeaders, "include", "i", false, "Include HTTP response headers in the output")
 	cmd.Flags().BoolVar(&opts.Paginate, "paginate", false, "Make additional HTTP requests to fetch all pages of results")
 	cmd.Flags().StringVar(&opts.RequestInputFile, "input", "", "The file to use as body for the HTTP request")
+	cmd.Flags().BoolVar(&opts.Silent, "silent", false, "Silence the output")
```
Let's try to be more descriptive here: how does “Do not print response body" sound? The goal would be to describe what "silence" means in this case, since the word could be ambiguous

mislav reviewed (changes requested):
It's great that you're tackling this; thank you! I have a few asks 🙇

AliabbasMerchant committed:
`api --silent` Changes:
Show Response Headers (if requested) even with `--silent` flag
Shift silent tests to `Test_apiRun`
Changed usage string of `--silent` flag

mislav reviewed (approved):
Fantastic; thank you!

vilmibm reviewed (approved):
thanks!
